lib/pf/authentication.pm

```diff
@@ -243,19 +243,20 @@ sub authenticate {
     my $message;
     foreach my $current_source (@sources) {
         my $result;
+        my $connection;
         $logger->trace("Trying to authenticate '$username' with source '".$current_source->id."'");
         eval {
-            ($result, $message) = $current_source->authenticate($username, $password);
+            ($result, $message, $connection) = $current_source->authenticate($username, $password);
         };
         # First match wins!
         if ($result) {
             $logger->info("Authentication successful for $username in source ".$current_source->id." (".$current_source->type.")");
-            return ($result, $message, $current_source->id);
+            return ($result, $message, $current_source->id, $connection);
```

Comment:
I am not sure that I like adding another param to the return.

I would suggest the following : 
Instead of returning the source ID, we return the source and then we populate an attribute that we make the same across all sources for the connection (like _connection)

This will require rewriting the code around authentication calls but would make a cleaner job IMO.

Or maybe something else than what is above but its just that returning 4 elements is not really in our common practice